tests/utils/s3_bulk_renaming_test.py

```diff
@@ -0,0 +1,146 @@
+from unittest import mock
+
+import pytest
+
+from utils.s3.s3_bulk_renaming import (
+    detect_timestamps_without_minutes,
+    filter_objects,
+    list_objects,
+    rename_objects,
+    update_timestamp,
+)
+
+
+@pytest.fixture
+def sample_objects():
+    return [
+        {"Key": "file1.txt"},
+        {"Key": "file2.jpg"},
+        {"Key": "file3.txt"},
+        {"Key": "file4.txt"},
+        {
+            "Key": (
+                "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
+                "2023-06-06T16/ges/observation/1.1"
+            )
+        },
+        {
+            "Key": (
+                "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
+                "2023-06-06T16:00/ges/observation/1.1"
+            )
+        },
+    ]
+
+
+class TestDetectTimestamps:
+    def test_without_minutes(self):
+        key = (
+            "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
+            "2023-06-06T16/ges/observation/1.1"
+        )
+
+        assert key == detect_timestamps_without_minutes(key)
+
+    def test_with_minutes(self):
+        key = (
+            "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
+            "2023-06-06T16:00/ges/observation/1.1"
+        )
+
+        assert detect_timestamps_without_minutes(key) is None
+
+    def test_no_timestamp(self):
+        key = "foo/bar/baz"
+
+        assert detect_timestamps_without_minutes(key) is None
+
+
+class TestUpdateTimestamp:
+    def test_update_timestamp(self):
+        key = (
+            "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
+            "2023-06-06T16/ges/observation/1.1"
+        )
+        desired = (
+            "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
+            "2023-06-06T16:00/ges/observation/1.1"
+        )
+
+        assert desired == update_timestamp(key)
+
+    def test_simplified_timestamp(self):
+        key = "foo/2023-06-06T16/bar"
+        desired = "foo/2023-06-06T16:00/bar"
+
+        assert desired == update_timestamp(key)
+
+    def test_bare_timestamp(self):
+        key = "2023-06-06T16"
+        desired = "2023-06-06T16:00"
+
+        assert desired == update_timestamp(key)
+
+    def test_matching_timestamps(self):
+        key = "2023-06-06T16:00"
+        desired = "2023-06-06T16:00"
+
+        assert desired == update_timestamp(key)
+
+    def test_no_timestamp(self):
+        key = "foo/bar"
+        desired = "foo/bar"
+
+        assert desired == update_timestamp(key)
```

Comment:
Since these are all effectively the same test, I’d just write one test function and parametrize it:

```suggestion
@pytest.mark.parametrize(
    "desired,expected",
    (
        (
            (
                "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
                "2023-06-06T16/ges/observation/1.1"
            ),
            (
                "diagnostics.zarr/RTMA/WCOSS/CONUS/RRFS/REALTIME/uv/"
                "2023-06-06T16:00/ges/observation/1.1"
            ),
        ),
        ("foo/2023-06-06T16/bar", "foo/2023-06-06T16:00/bar"),
        ("2023-06-06T16", "2023-06-06T16:00"),
        ("2023-06-06T16:00", "2023-06-06T16:00"),
        ("foo/bar", "foo/bar"),
    )
)
def test_update_timestamp(desired, expected):
    assert desired == update_timestamp(key)
```

Reply:
Ah, that's a nice improvement! This test always felt a bit cludgy to me - I haven't played with parameterized tests too much yet but this is a great usecase.